pinot-core/src/main/java/com/linkedin/pinot/core/metadata/MetadataProvider.java

```diff
@@ -0,0 +1,35 @@
+/**
+ * Copyright (C) 2014-2018 LinkedIn Corp. (pinot-core@linkedin.com)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.linkedin.pinot.core.metadata;
+
+import com.linkedin.pinot.common.segment.SegmentMetadata;
+import java.io.File;
+
+
+/**
+ * The metadata provider will take an input zipped .tar.gz file and extract and return the pinot segment metadata.
+ * This class is used during segment upload to get the metadata we need to store in zk.
+ */
+public interface MetadataProvider {
+  /**
+   * Returns a SegmentMetadata object from a tarred file
+   * @param tarredSegmentFile
+   * @param unzippedSegmentDir
+   * @return
+   * @throws Exception
+   */
+  SegmentMetadata extractMetadata(File tarredSegmentFile, File unzippedSegmentDir) throws Exception;
```

Comment:
Do we expect more classes for this interface? What else may they be doing? Looks to me that this method is pretty specific and does not warrant an interface unless I am missing something. Is the unzippedSegmentDir expected to be used as a tempDir? Or, is it that the method uses ONE of tarred file or unzippedsegmentdir to extract metadata? Some javadocs are needed, esp. if this is to be an interface.

Reply:
I want the segment upload to be as clean as possible and that means that the whole chain will have the same behavior with different classes. The other implementation once we add metadata upload is for the MetadataProvider to simply unzip return the metadata, similar to a noop.

Reply:
In that case, perhaps rename 'tarredSegmentFile' to 'tarFile' and 'unzippedSegmentDir' to 'workingDir'?

Reply:
Suggest not getting class by name, but constructing the exact MetadataExtractor class based on the upload type. Suggest rename MetadataProvider to MeadataExftractor.